sql/core/src/main/scala/org/apache/spark/sql/execution/command/AlterTableCommandParser.scala

```diff
@@ -0,0 +1,428 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.command
+
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.catalyst.catalog.ExternalCatalog.TablePartitionSpec
+import org.apache.spark.sql.catalyst.expressions.{Ascending, Descending, SortDirection}
+import org.apache.spark.sql.catalyst.parser._
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.execution.datasources._
+import org.apache.spark.sql.types.StructType
+
+
+/**
+ * Helper object to parse alter table commands.
+ */
+object AlterTableCommandParser {
+  import ParserUtils._
+
+  /**
+   * Parse the given node assuming it is an alter table command.
+   */
+  def parse(node: ASTNode): LogicalPlan = {
+    node.children match {
+      case (tabName @ Token("TOK_TABNAME", _)) :: otherNodes =>
+        val tableIdent = extractTableIdent(tabName)
+        val partSpec = getClauseOption("TOK_PARTSPEC", node.children).map(parsePartitionSpec)
+        matchAlterTableCommands(node, otherNodes, tableIdent, partSpec)
+      case _ =>
+        parseFailed("Could not parse ALTER TABLE command", node)
+    }
+  }
+
+  private def cleanAndUnquoteString(s: String): String = {
+    cleanIdentifier(unquoteString(s))
+  }
+
+  /**
+   * Extract partition spec from the given [[ASTNode]] as a map, assuming it exists.
+   *
+   * Expected format:
+   *   +- TOK_PARTSPEC
+   *      :- TOK_PARTVAL
+   *      :  :- dt
+   *      :  +- '2008-08-08'
+   *      +- TOK_PARTVAL
+   *         :- country
+   *         +- 'us'
+   */
+  private def parsePartitionSpec(node: ASTNode): Map[String, String] = {
+    node match {
+      case Token("TOK_PARTSPEC", partitions) =>
+        partitions.map {
+          // Note: sometimes there's a "=", "<" or ">" between the key and the value
```

Comment:
Maybe it will be good to provide a concrete example? Is it caused by what a user type?